### ISSUE TITLE
CKEditor: alter image dialog

### ISSUE THREAD

**Issue by tsiegleauq** [OpenSlides/OpenSlides#2924]
Since in the future external image references in motions should not be supported anymore, therefore the CKEditor image dialog needs to be altered.

The current state of CKEditor is shown below:
![bildschirmfoto von 2017-02-01 12-12-49](https://cloud.githubusercontent.com/assets/10233032/22504699/f4d60e74-e877-11e6-9a29-c768505bcedc.png)

What we gonna need is the prohibition of everything that resembles an URL (also: the customer probably does not want or need to know what an URL is)

Also we want to select an image file form the OpenSlides File-System directly without entering the path of this image.
Furtermore, The layout and styling does not fit the rest of the OpenSlides styling.

- Remove everything URL related
- Add possibility to select image (images are stored as files)
- (optional) add an upload dialog
- alter the style to match the rest of OpenSlides

**Comment by matakuka** [comment]
I think the easiest way is:

1. removing the image plugin
2. Adding an own modified version of the plugin: On click of the button or on double click on an image, an own dialog is opened (angular dialog, ui like the rest of OpenSlides).
3. The own dialog has the images from rest/mediafiles/mediafile as dropdown, a preview, and a field for alt-text.

The difficulty I am seeing is the "disallow external images" part. Currently there seems to be no mechanism to disallow elements based on url in CKeditor. 
We could do this on save, checking for (`location + '/media/file/'`), and change those into relative urls (openSlides' 'location' may change until next session).

This may confuse users: You are able to enter external images (e.g. via source code), but then it is removed again at save time.

**Comment by FinnStutzenstein** [comment]
@matakuka Just one thought: Can you open a ngDialog over an existing ngDialog?

**Comment by matakuka** [comment]
@FinnStutzenstein. Thanks for the remark. Apparently, this is not easy. As an alternative, I could use the ckeditor 'widget', but this won't look Openslides-ish.

**Comment by tsiegleauq** [comment]
> The difficulty I am seeing is the "disallow external images" part. Currently there seems to be no mechanism to disallow elements based on url in CKeditor.

usually, you should not expect the user to fiddle around with URLs, so you only offer the images which are already in OpenSlides filesystem (present a list of images, etc.) and nothing from external sources.

We might also include an option to directly upload/drag'n'drop-load an image over the new dialog.

**Comment by matakuka** [comment]
Yes, I understood that the following should be the "default" approach: A dialog with a drop-down list of images on the database, perhaps a preview of the selected image.

I'm thinking of "what if an (advanced) user tries to insert external images which we do not allow?"

Drag and drop is yet another thing.

**Comment by tsiegleauq** [comment]
The UI should not offer an easy way to enter any URL what so ever.

Tough, you can still manually edit the sourcecode of the motion, which allows you to do stuff like
`<img src="external.jpg">` etc.

before saving a motion, we should therefore parse all the image tags. Extracting this is not too hard. We can use a regex function like this to check if the given URL is from an external server or not:

```
function isExternal(url) {
    var match = url.match(/^([^:\/?#]+:)?(?:\/\/([^\/?#]*))?([^?#]+)?(\?[^#]*)?(#.*)?/);
    if (typeof match[1] === "string" && match[1].length > 0 && match[1].toLowerCase() !== location.protocol) return true;
    if (typeof match[2] === "string" && match[2].length > 0 && match[2].replace(new RegExp(":("+{"http:":80,"https:":443}[location.protocol]+")?$"), "") !== location.host) return true;
    return false;
}
```

**Comment by matakuka** [comment]
Apparently, we are not understanding each other, because that's more or less exactly what I was/am saying.

**Comment by donovaly** [comment]
This bug is in my opinion a MUSTFIX for 2.1 not 2.2. Currently it is not possible to display images in agenda texts. For example when you don't have Internet access during the assembly also a currently allowed external URL does not help.
I don't know technical details here, I just ant to fight for a method to display local images again in agenda texts as this was  possible before.

**Comment by emanuelschuetze** [comment]
Workaround how you can show an uploaded image in a topic (agenda item) slide:
* Upload an image file via OpenSlides' "Files" main menu
* Mouse over the image title and copy link path (e.g. http://localhost:8000/media/file/image.jpg)
* Create new agena item
* Click on "image" button in CKEditor toolbar (image url dialog is opened)
* Paste the copied (local, absolute) image url from your uploaded image file (e.g. `/media/file/image.jpg` is enough)
* Change some image properties if you like (e.g. width/height)
* Save agenda item and project slide

This works successfully with current master. There is no reason that a missing image list feature is a blocker for 2.1. We will add the feature after plugin evaluation for 2.2.

**Comment by donovaly** [comment]
> Mouse over the image title and copy link path (e.g. http://localhost:8000/media/file/image.jpg)

That works. Many thanks.
However this is not quick but dirty. I doubt that average OS users will come to this idea. At least I did not.
Please tell me if this workaround will definitely be the solution for OS 2.1 and I will add this to the documentation.

**Comment by donovaly** [comment]
My proposal for a better workaround: 
a field is added to the image insert dialog where one can click and the file open dialog of my operating system pops up letting me select a local file (the same as when pressing the "Durchsuchen" button when uploading a file to OS). This file will automatically be uploaded to OpenSlides and automatically converted into a http://localhost:8000 link.

By the way, the current URL input field cannot handle local links. For example the URL to my local file is
file:///D:/OpenSlides/Docs/de/clipart/Projektorsymbol-erweitert.jpg
Using this URL in my local browser displays the image correctly. But I cannot use this URL in the image insert dialog.

**Comment by tsiegleauq** [comment]
> However this is not quick but dirty. I doubt that average OS users will come to this idea.

I absolutely agree. That's why I created this issue. From a UI/UX engineering perspective, the current situation is unusable.

**However**:
> This bug is in my opinion a MUSTFIX for 2.1 not 2.2

This is (still) a feature request (and therefor not really a bug). Currently we stick to the CKEditors implementation afaik.

**Comment by emanuelschuetze** [comment]
This feature is alread implemented with #3337 and will be released with 2.2.

**Comment by donovaly** [comment]
Many thanks! Works fine, there is only this icon issue: #3553
